### PR TITLE
[UI] Make sure form validation displays non-valid fields as red in all forms

### LINKED ISSUE #7054: [UI] Form validation should display non-valid fields as Red in all forms
image

- [ ] Volume Mounts
- [ ] Environment variables
- [ ] Command and Args
- [ ] Deployment and Service Annotations
- [ ] Image Build args
- [ ] Command Selector (when creating a composite command)
- [ ] Container Selector (when creating an exec command)
- [ ] Image Selector (when creating an image command)
- [ ] Resource Selector (when creating an apply command)

### PR DESCRIPTION
**Pull request by rm3l**
**What type of PR is this:**
/area Web-UI

**What does this PR do / why we need it:**
This makes sure invalid fields are highlighted as red in all forms.

<details>
<summary>Add an Exec Command</summary>

image

</details>

<details>
<summary>Add an Image Command</summary>

image

</details>

<details>
<summary>Add an Apply Command</summary>

image

</details>

<details>
<summary>Add a Composite Command</summary>

image

</details>

<details>
<summary>Add a new container</summary>

image

image

</details>

<details>
<summary>Add a new image</summary>

image

</details>

<details>
<summary>Add a new resource</summary>

image

</details>

<details>
<summary>Add a new volume</summary>

image

</details>

**Which issue(s) this PR fixes:**
Fixes #7054 

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: ui/src/app/controls/endpoints/endpoints.component.ts [hunk 1]
```diff
@@ -30,20 +30,23 @@ export class EndpointsComponent implements ControlValueAccessor, Validator {
   form = new FormArray<FormGroup>([]);
 
   constructor() {
+    this.form.markAllAsTouched();
```
Comment (feloy) [inline ui/src/app/controls/endpoints/endpoints.component.ts:33]: If I understand correctly, you are adding this to make all the fields with validation marked as red from the beginning. I don't think it is idiomatic to Material Design, where required fields are marked with (*), and validation of fields with specific pattern are considered when you are typing into them only

Comment (rm3l) [reply ui/src/app/controls/endpoints/endpoints.component.ts:33]: That was initially a temporary way for me to quickly see at a glance if validation worked correctly and which fields were invalid or required.
But indeed, you're right. From the material design guidelines: `Show error text only after user interaction with a field`.
I'll revert the related commit soon.
Thanks.

**Review by rm3l — COMMENTED** [review]
@feloy As discussed together, I've made the necessary changes to simplify the form control logic. Please take a look when you get a chance. Thanks.

File: ui/src/app/forms/volume/volume.component.html [hunk 2]
```diff
@@ -10,6 +10,7 @@ <h2 *ngIf="volume">Edit volume <i>{{volume.name}}</i></h2>
         </mat-form-field>
         <mat-form-field appearance="outline" class="mid-width">
             <mat-label><span>Size</span></mat-label>
+            <mat-error>Example of valid quantities: 3, 3000M, 3G, 3Gi</mat-error>
```
Comment (feloy) [inline ui/src/app/forms/volume/volume.component.html:13]: The user may think these 4 quantities are equivalent (especially the "3" without any unit). 
I would give more information on the value, for example:

`300k (300*1000), 30Mi(30*1024²), 3Gi (3*1024³), 3G (3*1000³)`

File: ui/cypress/e2e/spec.cy.ts [hunk 3]
```diff
@@ -62,15 +62,15 @@ describe('devfile editor spec', () => {
     cy.getByDataCy('container-env-value-2').type("val3");
 
     cy.getByDataCy('volume-mount-add').click();
-    cy.getByDataCy('volume-mount-path-0').type("/mnt/vol1");
+    cy.getByDataCy('volume-mount-path-0').type("/mnt/vol1", {force: true});
```
Comment (feloy) [inline ui/cypress/e2e/spec.cy.ts:65]: Why is this necessary?

File: ui/src/app/forms/volume/volume.component.html (hunk 2 above)
Comment (rm3l) [reply ui/src/app/forms/volume/volume.component.html]: Makes sense. Fixed in [`9177d76` (#7064)](https://github.com/redhat-developer/odo/pull/7064/commits/9177d763ad898f7a3f68b9f3ab57c5b38c5940a5)

File: ui/cypress/e2e/spec.cy.ts (hunk 3 above)
Comment (rm3l) [inline ui/cypress/e2e/spec.cy.ts:65]: Not sure why, but I had to make Cypress forcibly bypass some checks with `force: true`. Otherwise, Cypress would complain that the element was covered by another one:

```
CypressError: Timed out retrying after 4000ms: `cy.type()` failed because this element:

`<input _ngcontent-xcd-c181="" formcontrolname="path" matinput="" class="mat-mdc-input-element ng-tns-c84-53 ng-untouched ng-pristine ng-invalid mat-mdc-form-field-input-control mdc-text-field__input cdk-text-field-autofill-monitored" ng-reflect-name="path" data-cy="volume-mount-path-0" id="mat-input-32" required="" aria-required="true">`

is being covered by another element:

`<span aria-hidden="true" class="mat-mdc-form-field-required-marker mdc-floating-label--required ng-tns-c84-53 ng-star-inserted"></span>`

Fix this problem, or use {force: true} to disable error checking.
```

**Review by feloy — APPROVED** [review]

**Comment by rm3l** [comment]
```
  [FAILED] in [It] - /go/odo_1/tests/helper/helper_cmd_wrapper.go:120 @ 09/05/23 12:54:55.574
  Running kubectl with args [kubectl get namespace cmd-devfile-deploy-test695stb -o jsonpath={.metadata.name}] and odo env: []
  [kubectl] cmd-devfile-deploy-test695stbDeleting project: cmd-devfile-deploy-test695stb
  Running kubectl with args [kubectl delete namespaces cmd-devfile-deploy-test695stb --wait=false] and odo env: []
  [kubectl] namespace "cmd-devfile-deploy-test695stb" deleted
  Setting current dir to: /go/odo_1/tests/integration
  Deleting dir: /tmp/3472070801
  Deleting dir: /tmp/2108200904
  << Timeline

  [FAILED] Timed out after 600.000s.
  Expected process to exit.  It did not.
  In [It] at: /go/odo_1/tests/helper/helper_cmd_wrapper.go:120 @ 09/05/23 12:54:55.574
...
Summarizing 1 Failure:
  [FAIL] odo devfile deploy command tests deploying devfile with long-running exec when Automount volumes are present in the namespace [It] should mount the volumes
  /go/odo_1/tests/helper/helper_cmd_wrapper.go:120

Ran 409 of 954 Specs in 1327.923 seconds
FAIL! -- 408 Passed | 1 Failed | 0 Pending | 545 Skipped
```

UI-only PR, so not related to the failures here.

/override Kubernetes-Integration-Tests/Kubernetes-Integration-Tests

**Comment by rm3l** [comment]
```
       âš   0/3 nodes are available: 3 pod has unbound immediate PersistentVolumeClaims. preemption: 0/3 nodes are available: 3 Preemption is not helpful for scheduling.
       âš   Pod is Pending
       âš   0/3 nodes are available: 3 pod has unbound immediate PersistentVolumeClaims. preemption: 0/3 nodes are available: 3 Preemption is not helpful for scheduling.
       âš   AttachVolume.Attach failed for volume "pvc-1606fb7f-4869-4b2a-b0d5-9fd061018b19" : timed out waiting for external-attacher of vpc.block.csi.ibm.io CSI driver to attach volume r010-ead57323-f582-4fee-a68b-c7fcbd5c5092
       âš   AttachVolume.Attach failed for volume "pvc-8521c60b-5aa0-433d-86dc-8bc09d2ba498" : timed out waiting for external-attacher of vpc.block.csi.ibm.io CSI driver to attach volume r010-cafd8507-9798-4a46-8a0f-440ba39d1c04
       âš   Unable to attach or mount volumes: unmounted volumes=[firstvol-jwjpal-app-vol secondvol-jwjpal-app-vol], unattached volumes=[odo-shared-data firstvol-jwjpal-app-vol secondvol-jwjpal-app-vol kube-api-access-xtxx6 odo-projects]: timed out waiting for the condition

...
Summarizing 3 Failures:
  [FAIL] odo devfile deploy command tests deploying devfile with long-running exec when Automount volumes are present in the namespace [It] should mount the volumes
  /go/odo_1/tests/helper/helper_cmd_wrapper.go:120
  [FAIL] odo dev command tests when running odo dev with devfile containing volume-component - without metadata.name [BeforeEach] should successfully use the volume components in container components
  /go/odo_1/tests/helper/helper_run.go:54
  [FAIL] odo dev command tests when running odo dev with devfile containing volume-component - with metadata.name [BeforeEach] should successfully use the volume components in container components
  /go/odo_1/tests/helper/helper_run.go:54

Ran 404 of 954 Specs in 1598.498 seconds
FAIL! -- 401 Passed | 3 Failed | 0 Pending | 550 Skipped
```

Looks like there are some issues with volumes. Let's keep an eye out for those failures to see if they happen again.
Overriding because this is a UI-only PR, so not related to the failures here.

/override OpenShift-Integration-tests/OpenShift-Integration-tests

**Comment by rm3l** [comment]
```

  [FAILED] Timed out after 420.012s.
  Expected
      <string>:   __
       /  \__     Developing using the "wmrbzj" Devfile
       \__/  \    Namespace: cmd-dev-test577yhx
       /  \__/    odo version: v3.14.0
       \__/
      
      - Running on the cluster in Dev mode
       -  Waiting for Kubernetes resources  ...
       !  Pod is Pending
       !  0/3 nodes are available: 3 pod has unbound immediate PersistentVolumeClaims. preemption: 0/3 nodes are available: 3 Preemption is not helpful for scheduling.
       !  0/3 nodes are available: 3 pod has unbound immediate PersistentVolumeClaims. preemption: 0/3 nodes are available: 3 Preemption is not helpful for scheduling.
       !  AttachVolume.Attach failed for volume "pvc-0e9d83df-b6c5-479c-b131-bbe589324d8a" : timed out waiting for external-attacher of vpc.block.csi.ibm.io CSI driver to attach volume r010-75cd0332-23ad-484f-b11d-7c4827c29cf6
       !  Unable to attach or mount volumes: unmounted volumes=[odo-projects], unattached volumes=[odo-projects odo-shared-data kube-api-access-mphwf]: timed out waiting for the condition
      
  to contain substring
      <string>: [Ctrl+c] - Exit
  In [BeforeEach] at: C:/Users/Administrator.ANSIBLE-TEST-VS/4557/tests/helper/helper_run.go:54 @ 09/05/23 07:54:40.296
[...]

Summarizing 5 Failures:
  [FAIL] odo dev command tests when a component is bootstrapped when odo dev is executed and Ephemeral is set to false [BeforeEach] should have created resources
  C:/Users/Administrator.ANSIBLE-TEST-VS/4557/tests/helper/helper_run.go:54
  [FAIL] odo dev command tests when a component is bootstrapped when odo dev is executed and Ephemeral is set to false [BeforeEach] when stopping odo dev normally should have deleted all resources before returning
  C:/Users/Administrator.ANSIBLE-TEST-VS/4557/tests/helper/helper_run.go:54
  [FAIL] odo dev command tests when a component is bootstrapped when odo dev is executed and Ephemeral is set to false [BeforeEach] when killing odo dev and running odo delete component --wait should have deleted all resources before returning
  C:/Users/Administrator.ANSIBLE-TEST-VS/4557/tests/helper/helper_run.go:54
  [FAIL] odo dev command tests when running odo dev with devfile containing volume-component - with metadata.name [BeforeEach] should successfully use the volume components in container components
  C:/Users/Administrator.ANSIBLE-TEST-VS/4557/tests/helper/helper_run.go:54
  [FAIL] odo dev command tests when running odo dev with devfile containing volume-component - without metadata.name [BeforeEach] should successfully use the volume components in container components
  C:/Users/Administrator.ANSIBLE-TEST-VS/4557/tests/helper/helper_run.go:54

Ran 388 of 954 Specs in 1476.036 seconds
FAIL! -- 383 Passed | 5 Failed | 0 Pending | 566 Skipped
```

Looks like there are some issues with volumes. Let's keep an eye out for those failures to see if they happen again.
Overriding because this is a UI-only PR, so not related to the failures here.

/override windows-integration-test/Windows-test
